odbc/src/api.rs

```diff
@@ -7,11 +11,49 @@ use odbc_sys::{
 
 #[no_mangle]
 pub extern "C" fn SQLAllocHandle(
-    _handle_type: HandleType,
-    _input_handle: Handle,
-    _output_handle: *mut Handle,
-) -> SqlReturn {
-    unimplemented!()
+    handle_type: HandleType,
+    input_handle: Handle,
+    output_handle: *mut Handle,
+) -> SqlReturn {
+    match handle_type {
+        HandleType::Env => {
+            let env = Box::new(EnvHandle::new());
+            env.env.write().unwrap().state = EnvState::Allocated;
+            unsafe { *output_handle = Box::into_raw(env) as *mut _ }
+        }
+        HandleType::Dbc => {
+            let env = input_handle as *mut _;
+            let conn = Box::new(ConnectionHandle::new(env));
+            conn.connection.write().unwrap().state =
+                ConnectionState::AllocatedEnvAllocatedConnection;
+            unsafe {
+                let conn_ptr = Box::into_raw(conn) as *mut _;
+                (*env).env.write().unwrap().connections.insert(conn_ptr);
```

Comment:
Note these `unwraps` are safe and what is commonly done with locks. They can only be errors if another thread holding the lock panics.

Reply:
In the MVP I foolishly did busy waiting on these, but they actually already sleep until the lock is available.